File: Data Structures & Algorithms/count-paths/submission-2.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        helper = [0]*(n + 1)
        helper[n - 1] = 1
        for i in range(m - 1, -1, -1):
            row = [0]*(n + 1)
            for j in range(n - 1, -1, -1):
                row[j] = helper[j] + row[j + 1]
            helper = row

        return helper[0]

File: Data Structures & Algorithms/combinations-of-a-phone-number/submission-2.cpp
class Solution {
private:
    string codes[10] = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        
    vector<string> combinations;
    string combination;

    void helper(int pos, string digits){
        if (pos == 0)combination = "";
        if (pos == digits.size()){
            combinations.push_back(combination);
            return;
        }
        int digit = digits[pos] - '0';
        for (char ch: codes[digit]){
            combination += ch;
            helper(pos + 1, digits);
            combination.pop_back();
        }
    }
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0)return combinations;
        helper(0, digits);
        return combinations;
    }
};
